### DISCUSSION TITLE
`git push origin HEAD:main` as a jj one-liner?

### DISCUSSION THREAD

**Discussion by wez** [jj-vcs/jj#3135] [Q&A]
I'm used to a completely branchless mercurial workflow, so I'm happy to ignore and do away with all local branches. I also happen to work primarily with git repos where I am the primary committer and I tend to push directly to the `main` branch on github, so I don't usually make PRs for my own work.  In git I use `git push origin HEAD:main` when I've got a stack of commits to push up.

I've only used `jj` to push a handful of commits so far, and each time I've stumbled with the need to run `jj branch set main -r @-` first. I would like to collapse that step out of my workflow.

From the `jj git push` docs it looks like I should be using `-c @-` (or `-r @-`?) to push the parent of my working copy, but the docs say that it will create a new branch for that. How can I tell it to push to `main@origin` and not make a new branch?

Is there a different way to express this workflow in `jj` that I've completely missed?

(Slight aside: `jj git push --help` docs feel a bit ambiguous regarding local vs. remote branches. I recognize that I'm new to `jj` and that perhaps it is more obvious to an established `jj` user, but I wanted to share that it isn't clear to me whether the branch creation is local or remote or how those interact around the push operation, from reading the `jj git push --help` docs in isolation)

**Comment by martinvonz** [comment]
You're correct that this is a missing feature. It doesn't seem to be a good fit for `jj git push` because that command pushes local branch to a remote without renaming them. We have talked about creating a separate command for pushing a revision to an arbitrary ref on the remote.

**Comment by arxanas** [comment]
If your repository is colocated, you should also be able to run `git push` directly.

**Reply by JumpIn-Git** [reply]
this doesnt seem to work for me...

**Reply by martinvonz** [reply]
If your repository really is [colocated](https://jj-vcs.github.io/jj/latest/git-compatibility/#co-located-jujutsugit-repos), then that's a question for a Git forum to answer why `git push` doesn't work.

**Comment by evmar** [comment]
Found this because I was facing the same.

One-liner for colocated is

```
$ git push origin HEAD:main
```

(and note that also updates branches like `main@origin` such that jj updates its local branches appropriately)

**Comment by jennings** [comment]
The equivalent of `git push origin HEAD:main` is `jj git push --named main=@`

**Reply by andy47** [reply]
Hi, that doesn't seem to work for me, I get an error message "error: unexpected argument '--named' found". I'm running jj 0.23, could this be in a newer version?

**Reply by PhilipMetzger** [reply]
yes, this flag was introduced in 0.29

**Reply by andy47** [reply]
Thanks, I'm not an expert in git or jj so apologies for the simple questions. I've just upgraded to 0.32.0 and when I try `jj git push --named main=@` now I get this message; 

    Error: Bookmark already exists: main
    Hint: Use 'jj bookmark move' to move it, and 'jj git push -b main [--allow-new]' to push it

**Reply by PhilipMetzger** [reply]
> Thanks, I'm not an expert in git or jj so apologies for the simple questions. I've just upgraded to 0.32.0 and when I try `jj git push --named main=@` now I get this message;

It fails because it always expects to create a new bookmark, which isn't the case for an already tracked bookmark like `main`. So you'd need to push it with `jj git push -b main`.

**Comment by vikks** [comment]
For me moving the bookmark to `@` works.
```shell
jj bookmark move main
jj git push -b main
```
